sr_port/gvcst_reversequery.c

```diff
@@ -0,0 +1,225 @@
+#include "jnl.h"		/* needed for tp.h */
+#include "gdskill.h"		/* needed for tp.h */
+#include "buddy_list.h"		/* needed for tp.h */
+#include "hashtab_int4.h"	/* needed for tp.h */
+#include "tp.h"			/* needed for T_BEGIN_READ_NONTP_OR_TP macro */
+#ifdef UNIX			/* needed for frame_pointer in GVCST_ROOT_SEARCH_AND_PREP macro */
+# include "repl_msg.h"
+# include "gtmsource.h"
+# include "rtnhdr.h"
+# include "stack_frame.h"
+# include "wbox_test_init.h"
+#endif
+
+#include "t_end.h"
+#include "t_retry.h"
+#include "t_begin.h"
+#include "gvcst_expand_key.h"
+#include "gvcst_protos.h"	/* for gvcst_rtsib,gvcst_search,gvcst_search_blk,gvcst_reversequery prototype */
+
+/* needed for spanning nodes */
+#include "op.h"
+#include "op_tcommit.h"
+#include "error.h"
+#include "tp_frame.h"
+#include "tp_restart.h"
+#include "gtmimagename.h"
+
+LITREF	mval		literal_batch;
+
+GBLREF sgmnt_addrs	*cs_addrs;
+GBLREF sgmnt_data_ptr_t	cs_data;
+GBLREF gd_region	*gv_cur_region;
+GBLREF gv_namehead	*gv_target;
+GBLREF gv_key		*gv_currkey, *gv_altkey;
+GBLREF uint4		dollar_tlevel;
+GBLREF unsigned int	t_tries;
+
+error_def(ERR_GVQUERYFAIL);
+
+DEFINE_NSB_CONDITION_HANDLER(gvcst_reversequery_ch)
+
+/* This function implements reverse $query for global nodes. With Standard NULL collation, a null subscript ("")
+ * is represented as the byte 0x01. Whereas a node that spans multiple GDS blocks (spanning node) has multiple nodes
+ * with hidden subscripts each of which start with the byte 0x02. So if $query(^x(1,2,3,7),-1) is requested, we
+ * might need to go back not just one node, but skip over hidden subscripts too until we find a node with non-hidden
+ * subscripts. Below is an example node layout where the subscripts of each node are listed. The result of the
+ * $query(^x(1,2,3,7)) operation should be ^x(1,2,3,"","",""). Hence the MAX_GVSUBSCRIPTS for loop below.
+ *
+ *  ^x(1,2,3,"")
+ *  ^x(1,2,3,"","")
+ *  ^x(1,2,3,"","","")              <--- reverse $query needs to end up here
+ *  ^x(1,2,3,"","","",hidden)
+ *  ^x(1,2,3,"","",hidden)
+ *  ^x(1,2,3,"",hidden)
+ *  ^x(1,2,3,hidden)
+ *  ^x(1,2,3,7)                     <--- reverse $query begins here
+ *
+ * "gv_currkey" already points to the input "gvn" for which reverse $query is sought.
+ * "gv_altkey" stores the result of the reverse $query at function end.
+ */
+boolean_t	gvcst_reversequery(void)
+{
+	boolean_t	found, is_hidden, sn_tpwrapped;
+	boolean_t	est_first_pass;
+	gv_key		save_currkey[DBKEYALLOC(MAX_KEY_SZ)];
+	int		i;
+	int		save_dollar_tlevel;
+
+	DEBUG_ONLY(save_dollar_tlevel = dollar_tlevel);
+	found = gvcst_reversequery2();
+	assert(save_dollar_tlevel == dollar_tlevel);
+	CHECK_HIDDEN_SUBSCRIPT_AND_RETURN(found, gv_altkey, is_hidden);
+	IF_SN_DISALLOWED_AND_NO_SPAN_IN_DB(return found);
+	assert(found && is_hidden);
+	SAVE_GV_CURRKEY(save_currkey);
+	if (!dollar_tlevel)
+	{
+		sn_tpwrapped = TRUE;
+		op_tstart((IMPLICIT_TSTART), TRUE, &literal_batch, 0);
```

Comment:
While it might not be intuitive, that is how the code needs to be unfortunately laid out. That is because the condition handler needs to take control inside a TP transaction so it has to be after the tstart. But restarts are possible even inside the tcommit. And so the REVERT has to happen after the tcommit. Added a comment to this effect in the code.